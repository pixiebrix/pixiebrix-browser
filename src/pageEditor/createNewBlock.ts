/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { BlockConfig } from "@/blocks/types";
import { defaultBlockConfig } from "@/blocks/util";
import { RegistryId, Schema } from "@/core";
import { uuidv4 } from "@/types/helpers";
import { getExampleBlockConfig } from "./exampleBlockConfigs";

export function createNewBlock(
  blockId: RegistryId,
  blockInputSchema?: Schema
): BlockConfig {
  return {
    id: blockId,
    instanceId: uuidv4(),
    config:
      getExampleBlockConfig(blockId) ??
      (blockInputSchema == null ? {} : defaultBlockConfig(blockInputSchema)),
  };
}
