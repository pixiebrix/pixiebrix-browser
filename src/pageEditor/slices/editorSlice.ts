/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { ActionFormState } from "@/pageEditor/extensionPoints/menuItem";
import { SidebarFormState } from "@/pageEditor/extensionPoints/sidebar";
import { TriggerFormState } from "@/pageEditor/extensionPoints/trigger";
import { PanelFormState } from "@/pageEditor/extensionPoints/panel";
import { ContextMenuFormState } from "@/pageEditor/extensionPoints/contextMenu";
import { getErrorMessage } from "@/errors";
import { clearExtensionTraces } from "@/telemetry/trace";
import { RegistryId, UUID } from "@/core";
import {
  ElementUIState,
  makeInitialElementUIState,
  makeInitialNodeUIState,
} from "@/pageEditor/uiState/uiState";
import {
  FOUNDATION_NODE_ID,
  NodeId,
} from "@/pageEditor/tabs/editTab/editorNodeLayout/EditorNodeLayout";
import { WritableDraft } from "immer/dist/types/types-external";
import { BlockConfig } from "@/blocks/types";
import { ExtensionPointType } from "@/extensionPoints/types";
import { OptionsDefinition, RecipeDefinition } from "@/types/definitions";

export type FormState =
  | ActionFormState
  | SidebarFormState
  | TriggerFormState
  | PanelFormState
  | ContextMenuFormState;

export interface EditorState {
  /**
   * A sequence number that changes whenever a new element is selected.
   *
   * Can use as a React component key to trigger a re-render
   */
  selectionSeq: number;

  /**
   * The element type, if the page editor is in "insertion-mode"
   */
  inserting: ExtensionPointType | null;

  /**
   * The uuid of the active element, if an extension is selected
   */
  activeElement: UUID | null;

  /**
   * The registry id of the active recipe, if a recipe is selected
   */
  activeRecipeId: RegistryId | null;

  error: string | null;

  dirty: Record<string, boolean>;

  /**
   * Unsaved elements
   */
  readonly elements: FormState[];

  /**
   * Brick ids (not UUIDs) that are known to be editable by the current user
   */
  knownEditable: RegistryId[];

  /**
   * True if error is because user does not have access to beta features
   */
  beta?: boolean;

  /**
   * Is the user using the new page editor beta UI?
   */
  isBetaUI: boolean;

  /**
   * The current UI state of each element, indexed by element Id
   */
  elementUIStates: Record<UUID, ElementUIState>;

  /**
   * A clipboard-style-copy of a block ready to paste into an extension
   */
  copiedBlock?: BlockConfig;

  /**
   * Are we currently showing the info message to users about upgrading from v2 to v3 of
   * the runtime api for this extension?
   */
  showV3UpgradeMessageByElement: Record<UUID, boolean>;

  /**
   * Unsaved, changed recipe options
   */
  dirtyRecipeOptionsById: Record<RegistryId, OptionsDefinition>;
}

export const initialState: EditorState = {
  selectionSeq: 0,
  activeElement: null,
  activeRecipeId: null,
  error: null,
  beta: false,
  elements: [],
  knownEditable: [],
  dirty: {},
  inserting: null,
  isBetaUI: false,
  elementUIStates: {},
  showV3UpgradeMessageByElement: {},
  dirtyRecipeOptionsById: {},
};

/* eslint-disable security/detect-object-injection, @typescript-eslint/no-dynamic-delete -- lots of immer-style code here dealing with Records */
function ensureElementUIState(
  state: WritableDraft<EditorState>,
  elementId: UUID
) {
  if (!state.elementUIStates[elementId]) {
    state.elementUIStates[elementId] = makeInitialElementUIState();
  }
}

function ensureNodeUIState(
  state: WritableDraft<ElementUIState>,
  nodeId: NodeId
) {
  if (!state.nodeUIStates[nodeId]) {
    state.nodeUIStates[nodeId] = makeInitialNodeUIState(nodeId);
  }
}

function syncElementNodeUIStates(
  state: WritableDraft<EditorState>,
  element: FormState
) {
  const elementUIState = state.elementUIStates[element.uuid];
  const blockPipelineIds = element.extension.blockPipeline.map(
    (x) => x.instanceId
  );

  // Pipeline block instance IDs may have changed
  if (!blockPipelineIds.includes(elementUIState.activeNodeId)) {
    elementUIState.activeNodeId = FOUNDATION_NODE_ID;
  }

  // Remove NodeUIStates for invalid IDs
  for (const key of Object.keys(elementUIState.nodeUIStates)) {
    const nodeId = key as NodeId;
    // Don't remove the foundation NodeUIState
    if (nodeId !== FOUNDATION_NODE_ID && !blockPipelineIds.includes(nodeId)) {
      delete elementUIState.nodeUIStates[nodeId];
    }
  }

  // Add missing NodeUIStates
  for (const uuid of blockPipelineIds) {
    ensureNodeUIState(elementUIState, uuid);
  }
}

function setActiveNodeId(state: WritableDraft<EditorState>, nodeId: NodeId) {
  const elementUIState = state.elementUIStates[state.activeElement];
  ensureNodeUIState(elementUIState, nodeId);
  elementUIState.activeNodeId = nodeId;
}

export const editorSlice = createSlice({
  name: "editor",
  initialState,
  reducers: {
    toggleInsert(state, action: PayloadAction<ExtensionPointType>) {
      state.inserting = action.payload;
      state.beta = false;
      state.error = null;
    },
    markEditable(state, action: PayloadAction<RegistryId>) {
      state.knownEditable.push(action.payload);
    },
    addElement(state, action: PayloadAction<FormState>) {
      const element = action.payload;
      state.inserting = null;
      state.elements.push(element);
      state.error = null;
      state.dirty[element.uuid] = true;
      state.beta = false;
      state.activeElement = element.uuid;
      state.selectionSeq++;
      state.elementUIStates[element.uuid] = makeInitialElementUIState();
    },
    betaError(state, action: PayloadAction<{ error: string }>) {
      state.error = action.payload.error;
      state.beta = true;
      state.activeElement = null;
    },
    adapterError(state, action: PayloadAction<{ uuid: UUID; error: unknown }>) {
      const { uuid, error } = action.payload;
      state.error = getErrorMessage(error);
      state.beta = false;
      state.activeElement = uuid;
      state.selectionSeq++;
    },
    selectInstalled(state, action: PayloadAction<FormState>) {
      const { uuid } = action.payload;
      const index = state.elements.findIndex((x) => x.uuid === uuid);
      if (index >= 0) {
        state.elements[index] = action.payload;
      } else {
        state.elements.push(action.payload);
      }

      state.error = null;
      state.beta = null;
      state.activeElement = uuid;
      state.activeRecipeId = null;
      state.selectionSeq++;
      ensureElementUIState(state, uuid);
    },
    resetInstalled(state, actions: PayloadAction<FormState>) {
      const element = actions.payload;
      const index = state.elements.findIndex((x) => x.uuid === element.uuid);
      if (index >= 0) {
        state.elements[index] = element;
      } else {
        state.elements.push(element);
      }

      state.dirty[element.uuid] = false;
      state.error = null;
      state.beta = null;
      state.selectionSeq++;

      // Make sure we're not keeping any private data around from Page Editor sessions
      void clearExtensionTraces(element.uuid);

      syncElementNodeUIStates(state, element);
    },
    selectElement(state, action: PayloadAction<UUID>) {
      const elementId = action.payload;
      if (!state.elements.some((x) => x.uuid === elementId)) {
        throw new Error(`Unknown dynamic element: ${action.payload}`);
      }

      state.error = null;
      state.beta = null;
      state.activeElement = elementId;
      state.activeRecipeId = null;
      state.selectionSeq++;
      ensureElementUIState(state, elementId);
    },
    markSaved(state, action: PayloadAction<UUID>) {
      const element = state.elements.find((x) => action.payload === x.uuid);
      if (!element) {
        throw new Error(`Unknown dynamic element: ${action.payload}`);
      }

      if (!element.installed) {
        state.knownEditable.push(element.extensionPoint.metadata.id);
      }

      element.installed = true;
      state.dirty[element.uuid] = false;
      // Force a reload so the _new flags are correct on the readers
      state.selectionSeq++;
    },
    /**
     * Sync the redux state with the form state.
     * Used on by the page editor to set changed version of the element in the store.
     */
    editElement(state, action: PayloadAction<FormState>) {
      const element = action.payload;
      const index = state.elements.findIndex((x) => x.uuid === element.uuid);
      if (index < 0) {
        throw new Error(`Unknown dynamic element: ${element.uuid}`);
      }

      state.elements[index] = element;
      state.dirty[element.uuid] = true;

      syncElementNodeUIStates(state, element);
    },
    /**
     * Applies the update to the element
     */
    updateElement(
      state,
      action: PayloadAction<{ uuid: UUID } & Partial<FormState>>
    ) {
      const { uuid, ...elementUpdate } = action.payload;
      const index = state.elements.findIndex((x) => x.uuid === uuid);
      if (index < 0) {
        throw new Error(`Unknown dynamic element: ${uuid}`);
      }

      // @ts-expect-error -- Concrete variants of FromState are not mutually assignable.
      state.elements[index] = {
        ...state.elements[index],
        ...elementUpdate,
      };

      // Force reload of Formik state
      state.selectionSeq++;
    },
    removeElement(state, action: PayloadAction<UUID>) {
      const uuid = action.payload;
      if (state.activeElement === uuid) {
        state.activeElement = null;
      }

      state.elements.splice(
        state.elements.findIndex((x) => x.uuid === uuid),
        1
      );

      delete state.dirty[uuid];
      delete state.elementUIStates[uuid];

      // Make sure we're not keeping any private data around from Page Editor sessions
      void clearExtensionTraces(uuid);
    },
    selectRecipe(state, action: PayloadAction<RecipeDefinition>) {
      const recipe = action.payload;
      state.error = null;
      state.beta = null;
      state.activeElement = null;
      state.activeRecipeId = recipe.metadata.id;
      state.selectionSeq++;
    },
    setBetaUIEnabled(state, action: PayloadAction<boolean>) {
      state.isBetaUI = action.payload;
    },
    removeElementNodeUIState(
      state,
      action: PayloadAction<{
        nodeIdToRemove: NodeId;
        newActiveNodeId?: NodeId;
      }>
    ) {
      const elementUIState = state.elementUIStates[state.activeElement];
      const { nodeIdToRemove, newActiveNodeId } = action.payload;

      const activeNodeId = newActiveNodeId ?? FOUNDATION_NODE_ID;
      setActiveNodeId(state, activeNodeId);

      delete elementUIState.nodeUIStates[nodeIdToRemove];
    },
    setElementActiveNodeId(state, action: PayloadAction<NodeId>) {
      setActiveNodeId(state, action.payload);
    },
    setNodeDataPanelTabSelected(state, action: PayloadAction<string>) {
      const elementUIState = state.elementUIStates[state.activeElement];
      const nodeUIState =
        elementUIState.nodeUIStates[elementUIState.activeNodeId];
      nodeUIState.dataPanel.activeTabKey = action.payload;
    },
    setNodeDataPanelTabSearchQuery(
      state,
      action: PayloadAction<{ tabKey: string; query: string }>
    ) {
      const { tabKey, query } = action.payload;
      const elementUIState = state.elementUIStates[state.activeElement];
      const nodeUIState =
        elementUIState.nodeUIStates[elementUIState.activeNodeId];
      nodeUIState.dataPanel.tabQueries[tabKey] = query;
    },
    copyBlockConfig(state, action: PayloadAction<BlockConfig>) {
      const copy = { ...action.payload };
      delete copy.instanceId;
      state.copiedBlock = copy;
    },
    clearCopiedBlockConfig(state) {
      delete state.copiedBlock;
    },
    showV3UpgradeMessage(state) {
      state.showV3UpgradeMessageByElement[state.activeElement] = true;
    },
    hideV3UpgradeMessage(state) {
      state.showV3UpgradeMessageByElement[state.activeElement] = false;
    },
    editRecipeOptions(state, action: PayloadAction<OptionsDefinition>) {
      const recipeId = state.activeRecipeId;
      if (recipeId == null) {
        return;
      }

      const { payload: options } = action;
      state.dirtyRecipeOptionsById[recipeId] = options;
    },
    resetRecipeOptions(state, action: PayloadAction<RegistryId>) {
      const { payload: recipeId } = action;
      delete state.dirtyRecipeOptionsById[recipeId];
    },
  },
});
/* eslint-enable security/detect-object-injection, @typescript-eslint/no-dynamic-delete -- re-enable rule */

export const { actions } = editorSlice;
