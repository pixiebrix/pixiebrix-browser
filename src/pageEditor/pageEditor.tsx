/*
 * Copyright (C) 2024 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import "@/vendors/bootstrapWithoutRem.css";
import "@/utils/global.scss";
import "@/utils/layout.scss";

import "@/extensionContext";
import "@/development/darkMode.js";

import { initMessengerLogging } from "@/development/messengerLogging";

import ReactDOM from "react-dom";
import React from "react";
import Panel from "@/pageEditor/Panel";
import { watchNavigation } from "@/pageEditor/protocol";
import { initToaster } from "@/utils/notify";
import { initRuntimeLogging } from "@/development/runtimeLogging";
import { initPerformanceMonitoring } from "@/telemetry/performance";
import { setPlatform } from "@/platform/platformContext";
import extensionPagePlatform from "@/extensionPages/extensionPagePlatform";

async function init() {
  setPlatform(extensionPagePlatform);
  void initMessengerLogging();
  void initRuntimeLogging();
  try {
    await initPerformanceMonitoring();
  } catch (error) {
    console.error("Failed to initialize performance monitoring", error);
  }

  watchNavigation();
  initToaster();

  ReactDOM.render(<Panel />, document.querySelector("#container"));
}

void init();
