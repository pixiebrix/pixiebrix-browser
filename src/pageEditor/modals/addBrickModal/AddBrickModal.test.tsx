/*
 * Copyright (C) 2024 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import { render, screen } from "@/pageEditor/testHelpers";
import AddBrickModal from "@/pageEditor/modals/addBrickModal/AddBrickModal";
import { actions } from "@/pageEditor/store/editor/editorSlice";
import userEvent from "@testing-library/user-event";
import { PIPELINE_BRICKS_FIELD_NAME } from "@/pageEditor/consts";
import { array } from "cooky-cutter";
import { waitForEffect } from "@/testUtils/testHelpers";
import brickRegistry from "@/bricks/registry";
import {
  echoBrick,
  featureFlagBrick,
} from "@/runtime/pipelineTests/testHelpers";
import { appApiMock } from "@/testUtils/appApiMock";
import { formStateFactory } from "@/testUtils/factories/pageEditorFactories";
import {
  marketplaceListingFactory,
  marketplaceTagFactory,
} from "@/testUtils/factories/marketplaceFactories";
import { PipelineFlavor } from "@/bricks/types";
import { API_PATHS } from "@/data/service/urlPaths";

// Need at least one item so callers see the registry as initialized
brickRegistry.register([echoBrick, featureFlagBrick]);

beforeAll(() => {
  const tags = array(marketplaceTagFactory, 3)({ subtype: "role" });
  const listings = array(marketplaceListingFactory, 10)({ tags });

  appApiMock.reset();
  appApiMock.onGet(API_PATHS.MARKETPLACE_TAGS).reply(200, tags);
  appApiMock.onGet(API_PATHS.MARKETPLACE_LISTINGS).reply(200, listings);
});

describe("AddBrickModal", () => {
  it("renders", async () => {
    const formState = formStateFactory();
    const { asFragment } = render(<AddBrickModal />, {
      setupRedux(dispatch) {
        dispatch(actions.addModComponentFormState(formState));
        dispatch(actions.setActiveModComponentId(formState.uuid));
        dispatch(
          actions.showAddBrickModal({
            path: "",
            flavor: PipelineFlavor.AllBricks,
            index: 0,
          }),
        );
      },
      // This currently produces a warning, but allows us to snapshot the modal
      //  See: https://github.com/testing-library/react-testing-library/issues/62
      container: document.body,
    });

    await waitForEffect();

    // Exclude the feature flagged brick by default
    expect(screen.getByText(echoBrick.name)).toBeInTheDocument();
    expect(screen.queryByText(featureFlagBrick.name)).not.toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });

  it("includes feature flagged brick", async () => {
    appApiMock.onGet(API_PATHS.FEATURE_FLAGS).reply(200, {
      flags: [featureFlagBrick.featureFlag],
    });

    const formState = formStateFactory();
    render(<AddBrickModal />, {
      setupRedux(dispatch) {
        dispatch(actions.addModComponentFormState(formState));
        dispatch(actions.setActiveModComponentId(formState.uuid));
        dispatch(
          actions.showAddBrickModal({
            path: "",
            flavor: PipelineFlavor.AllBricks,
            index: 0,
          }),
        );
      },
      // This currently produces a warning, but allows us to snapshot the modal
      //  See: https://github.com/testing-library/react-testing-library/issues/62
      container: document.body,
    });

    await waitForEffect();

    expect(screen.getByText(echoBrick.name)).toBeInTheDocument();
    expect(screen.getByText(featureFlagBrick.name)).toBeInTheDocument();
  });

  it("renders with tag selected and search query", async () => {
    const formState = formStateFactory();

    const { asFragment } = render(<AddBrickModal />, {
      setupRedux(dispatch) {
        dispatch(actions.addModComponentFormState(formState));
        dispatch(actions.setActiveModComponentId(formState.uuid));
        dispatch(
          actions.showAddBrickModal({
            path: PIPELINE_BRICKS_FIELD_NAME,
            flavor: PipelineFlavor.AllBricks,
            index: 0,
          }),
        );
      },
      container: document.body,
    });

    await waitForEffect();

    // Click the last tag
    const tags = screen.getAllByTestId("search-tag-item", {
      exact: false,
    });

    await userEvent.click(tags.at(-1)!);

    // Enter a query
    await userEvent.type(screen.getByTestId("tag-search-input"), "google");

    expect(asFragment()).toMatchSnapshot();
  });
});
