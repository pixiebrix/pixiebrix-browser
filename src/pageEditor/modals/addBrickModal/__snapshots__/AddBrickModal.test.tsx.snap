// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddBrickModal it renders 1`] = `
<DocumentFragment>
  <div
    class="fade modal-backdrop show"
  />
  <div
    aria-modal="true"
    class="fade root modal show"
    role="dialog"
    style="display: block;"
    tabindex="-1"
  >
    <div
      class="modal-dialog modal-xl modal-dialog-centered"
    >
      <div
        class="modal-content"
      >
        <div
          class="header modal-header"
        >
          <div
            class="title modal-title h4"
          >
            Add Brick
          </div>
          <div
            class="searchContainer"
          >
            <div
              class="root searchInput"
            >
              <input
                class="input"
                data-testid="tag-search-input"
                name="brickSearch"
                placeholder="Search"
                value=""
              />
            </div>
          </div>
          <button
            class="close closeButton"
            type="button"
          >
            <span
              aria-hidden="true"
            >
              Ã—
            </span>
            <span
              class="sr-only"
            >
              Close
            </span>
          </button>
        </div>
        <div
          class="body modal-body"
        >
          <div
            class="tagList"
          >
            <div
              class="root"
            >
              <button
                class="item itemActive"
                data-testid="search-tag-item-All Categories"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-tag fa-w-16 fa-fw "
                  data-icon="tag"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 252.118V48C0 21.49 21.49 0 48 0h204.118a48 48 0 0 1 33.941 14.059l211.882 211.882c18.745 18.745 18.745 49.137 0 67.882L293.823 497.941c-18.745 18.745-49.137 18.745-67.882 0L14.059 286.059A48 48 0 0 1 0 252.118zM112 64c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48z"
                    fill="currentColor"
                  />
                </svg>
                 All Categories
              </button>
              <button
                class="item"
                data-testid="search-tag-item-Tag 1"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-tag fa-w-16 fa-fw "
                  data-icon="tag"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 252.118V48C0 21.49 21.49 0 48 0h204.118a48 48 0 0 1 33.941 14.059l211.882 211.882c18.745 18.745 18.745 49.137 0 67.882L293.823 497.941c-18.745 18.745-49.137 18.745-67.882 0L14.059 286.059A48 48 0 0 1 0 252.118zM112 64c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48z"
                    fill="currentColor"
                  />
                </svg>
                 Tag 1
              </button>
              <button
                class="item"
                data-testid="search-tag-item-Tag 2"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-tag fa-w-16 fa-fw "
                  data-icon="tag"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 252.118V48C0 21.49 21.49 0 48 0h204.118a48 48 0 0 1 33.941 14.059l211.882 211.882c18.745 18.745 18.745 49.137 0 67.882L293.823 497.941c-18.745 18.745-49.137 18.745-67.882 0L14.059 286.059A48 48 0 0 1 0 252.118zM112 64c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48z"
                    fill="currentColor"
                  />
                </svg>
                 Tag 2
              </button>
              <button
                class="item"
                data-testid="search-tag-item-Tag 3"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-tag fa-w-16 fa-fw "
                  data-icon="tag"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 252.118V48C0 21.49 21.49 0 48 0h204.118a48 48 0 0 1 33.941 14.059l211.882 211.882c18.745 18.745 18.745 49.137 0 67.882L293.823 497.941c-18.745 18.745-49.137 18.745-67.882 0L14.059 286.059A48 48 0 0 1 0 252.118zM112 64c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48z"
                    fill="currentColor"
                  />
                </svg>
                 Tag 3
              </button>
            </div>
          </div>
          <div
            class="brickResults"
          >
            <div
              style="position: relative; height: 600px; width: 600px; overflow: auto; will-change: transform; direction: ltr;"
            >
              <div
                style="height: 89px; width: 600px;"
              >
                <div
                  style="position: absolute; left: 0px; top: 0px; height: 89px; width: 300px;"
                >
                  <div
                    class="root"
                    role="button"
                    tabindex="0"
                  >
                    <div
                      class="content"
                    >
                      <div
                        class="nameRow"
                      >
                        <img
                          alt="Icon"
                          class="imageIcon size1"
                          src="https://marketplace.dev/1"
                        />
                        <span
                          class="name"
                        >
                          Echo Brick
                        </span>
                      </div>
                      <small
                        class="text-muted font-italic"
                      >
                        No description provided.
                      </small>
                    </div>
                    <button
                      class="addButton btn btn-primary"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="svg-inline--fa fa-plus fa-w-14 "
                        data-icon="plus"
                        data-prefix="fas"
                        focusable="false"
                        role="img"
                        viewBox="0 0 448 512"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                          fill="currentColor"
                        />
                      </svg>
                       Add
                    </button>
                  </div>
                </div>
                <div
                  style="position: absolute; left: 300px; top: 0px; height: 89px; width: 300px;"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`AddBrickModal it renders with tag selected and search query 1`] = `
<DocumentFragment>
  <div
    class="fade modal-backdrop show"
  />
  <div
    aria-modal="true"
    class="fade root modal show"
    role="dialog"
    style="display: block;"
    tabindex="-1"
  >
    <div
      class="modal-dialog modal-xl modal-dialog-centered"
    >
      <div
        class="modal-content"
      >
        <div
          class="header modal-header"
        >
          <div
            class="title modal-title h4"
          >
            Add Brick
          </div>
          <div
            class="searchContainer"
          >
            <div
              class="root searchInput"
            >
              <span
                class="badge"
              >
                Tag 3 
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-times-circle fa-w-16 "
                  data-icon="times-circle"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z"
                    fill="currentColor"
                  />
                </svg>
              </span>
              <input
                class="input"
                data-testid="tag-search-input"
                name="brickSearch"
                placeholder="Search"
                value="google"
              />
            </div>
          </div>
          <button
            class="close closeButton"
            type="button"
          >
            <span
              aria-hidden="true"
            >
              Ã—
            </span>
            <span
              class="sr-only"
            >
              Close
            </span>
          </button>
        </div>
        <div
          class="body modal-body"
        >
          <div
            class="tagList"
          >
            <div
              class="root"
            >
              <button
                class="item"
                data-testid="search-tag-item-All Categories"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-tag fa-w-16 fa-fw "
                  data-icon="tag"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 252.118V48C0 21.49 21.49 0 48 0h204.118a48 48 0 0 1 33.941 14.059l211.882 211.882c18.745 18.745 18.745 49.137 0 67.882L293.823 497.941c-18.745 18.745-49.137 18.745-67.882 0L14.059 286.059A48 48 0 0 1 0 252.118zM112 64c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48z"
                    fill="currentColor"
                  />
                </svg>
                 All Categories
              </button>
              <button
                class="item"
                data-testid="search-tag-item-Tag 1"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-tag fa-w-16 fa-fw "
                  data-icon="tag"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 252.118V48C0 21.49 21.49 0 48 0h204.118a48 48 0 0 1 33.941 14.059l211.882 211.882c18.745 18.745 18.745 49.137 0 67.882L293.823 497.941c-18.745 18.745-49.137 18.745-67.882 0L14.059 286.059A48 48 0 0 1 0 252.118zM112 64c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48z"
                    fill="currentColor"
                  />
                </svg>
                 Tag 1
              </button>
              <button
                class="item"
                data-testid="search-tag-item-Tag 2"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-tag fa-w-16 fa-fw "
                  data-icon="tag"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 252.118V48C0 21.49 21.49 0 48 0h204.118a48 48 0 0 1 33.941 14.059l211.882 211.882c18.745 18.745 18.745 49.137 0 67.882L293.823 497.941c-18.745 18.745-49.137 18.745-67.882 0L14.059 286.059A48 48 0 0 1 0 252.118zM112 64c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48z"
                    fill="currentColor"
                  />
                </svg>
                 Tag 2
              </button>
              <button
                class="item itemActive"
                data-testid="search-tag-item-Tag 3"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-tag fa-w-16 fa-fw "
                  data-icon="tag"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 252.118V48C0 21.49 21.49 0 48 0h204.118a48 48 0 0 1 33.941 14.059l211.882 211.882c18.745 18.745 18.745 49.137 0 67.882L293.823 497.941c-18.745 18.745-49.137 18.745-67.882 0L14.059 286.059A48 48 0 0 1 0 252.118zM112 64c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48z"
                    fill="currentColor"
                  />
                </svg>
                 Tag 3
              </button>
            </div>
          </div>
          <div
            class="brickResults"
          >
            <div
              style="position: relative; height: 600px; width: 600px; overflow: auto; will-change: transform; direction: ltr;"
            >
              <div
                style="height: 0px; width: 600px;"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
